### PR TITLE
Add functionality to open the Settings UI tab through openSettings

### PR DESCRIPTION
**Pull request by leonMSFT**
This PR's goal was to add an option to the `OpenSettings` keybinding to open the Settings UI in a tab. In order to implement that, a couple of changes had to be made to `Tab`, specifically:

- Introduce a tab interface named `ITab`
- Create/Rename two new Tab classes that implement `ITab` called `SettingsTab` and `TerminalTab` 

I've also put some `TODOs` that I wanted to get some thoughts on - I'll make a follow up PR but perhaps they can be revisited when we flesh out the Settings UI more.
- Does a Settings UI tab shutdown need to do anything special for cleanup?
- Does a Settings UI tab need to have `GetActiveTitle`? Maybe depending on which page in the UI is open?
- Technically, I can't focus a `Grid` control, so I'll need to figure out what to `Focus` when the tab is selected.
- The Settings UI tab doesn't have a `TermControl`, so once focus is moved to that tab, users won't be able to `nextTab/prevTab` out of it (along with all other keybindings).

References #1564, #5915

### CODE REVIEW

File: src/cascadia/TerminalSettingsEditor/MainPage.idl
Comment (DHowett) [inline src/cascadia/TerminalSettingsEditor/MainPage.idl:7]: honestly page should be acceptable -- we may want to make the content property a FrameworkElement, not just a Grid or Control

File: src/cascadia/TerminalApp/ITab.idl
Comment (DHowett) [inline src/cascadia/TerminalApp/ITab.idl:15]: note below about FrameworkElement. We could use Page for the FE if we want?

File: src/cascadia/TerminalApp/ITab.idl
Comment (DHowett) [inline src/cascadia/TerminalApp/ITab.idl:16]: may wanna make this a property!

File: src/cascadia/TerminalApp/ITab.idl
Comment (DHowett) [inline src/cascadia/TerminalApp/ITab.idl:19]: may wanna make this a property! or, copy the `Focus`/`TryFocus` method from xaml UIElement/Control (!) and have the same signature

File: src/cascadia/TerminalApp/ITab.idl
Comment (DHowett) [inline src/cascadia/TerminalApp/ITab.idl:15]: make this a property, `Content`?

File: src/cascadia/TerminalApp/ITab.idl
Comment (DHowett) [inline src/cascadia/TerminalApp/ITab.idl:14]: property?

File: src/cascadia/TerminalApp/Utils.h
Comment (DHowett) [inline src/cascadia/TerminalApp/Utils.h:156]: neat! dedup this with Command's magic icon handler later.

**Review by DHowett — COMMENTED** [review]
note to self: review `SettingsTab`, `TerminalPage::_OpenSettingsUI`

File: src/cascadia/TerminalApp/ITab.idl
Comment (leonMSFT) [reply src/cascadia/TerminalApp/ITab.idl:16]: I've got a question around this one - would you say `TerminalTab` should consider `Title()` and `GetActiveTitle()` to return the same thing? When would we _not_ want the "active title"?

Comment (DHowett) [reply src/cascadia/TerminalApp/ITab.idl:16]: That’s a Q for @zadjii-msft! IDK why we would have separate titles

File: src/cascadia/TerminalApp/AppLogic.h
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/AppLogic.h:8]: Curious: What do we need this one for?

File: src/cascadia/TerminalApp/TerminalAppLib.vcxproj
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/TerminalAppLib.vcxproj:30]: nit: it'd be nice if we could keep the extra empty lines here and there. Makes it more legible.

File: src/cascadia/TerminalApp/TerminalPage.cpp
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/TerminalPage.cpp:2553]: doc comment

Comment (carlos-zamora) [reply src/cascadia/TerminalApp/TerminalPage.cpp:2553]: Specifically call out in the method comment that if the Settings UI is already open, we just switch to it.

File: src/cascadia/TerminalApp/TerminalPage.cpp
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/TerminalPage.cpp:2570]: Is this because we manually keep `_tabs` and `_tabView` in sync? :(

File: src/cascadia/TerminalApp/TerminalPage.cpp
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/TerminalPage.cpp:2581]: _urgh_. XAML2711 is closed; I guess we haven't ingested the change yet? Is it possible for us to consolidate this into a helper function? That way we just remove it altogether when we don't need it anymore.

File: src/cascadia/TerminalApp/TerminalPage.cpp
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/TerminalPage.cpp:2625]: Would this work?
```suggestion
        return _GetTerminalTabImpl(_tabs.GetAt(index));
```

Or could we remove this helper function altogether and just use the one below instead?

File: src/cascadia/TerminalApp/TerminalPage.h
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/TerminalPage.h:8]: wait... what do we need `SettingsTab.h` for? It looks like we're only working with `TerminalTab` below.

File: src/cascadia/TerminalApp/TerminalTab.h
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/TerminalTab.h:97]: I think I like this change. But why did we have to do it?

File: src/cascadia/TerminalApp/SettingsTab.h
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/SettingsTab.h:2]: Slap one of those doc comments with an abstract on here and put your name on it! :)

(this is the first one I noticed, but I'm almost at the end of my review. Could you check if you need to do the same with any of the other .h introduced too plz)

File: src/cascadia/TerminalApp/SettingsTab.cpp
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/SettingsTab.cpp:104]: Good thinking. This should probably be left as polish for after we get this into master though. Let's remove the comment for now and just add a note in the Epic that we should do this later.

File: src/cascadia/TerminalApp/SettingsTab.cpp
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/SettingsTab.cpp:113]: <details>
<summary>My train of thought 🚂🧠</summary>
When the Settings UI is opened, we to create a clone of CascadiaSettings.
As the user makes changes, we directly modify that clone.
When the user saves the changes, that clone gets promoted to be the real CascadiaSettings.
When we close the Settings UI, the Settings UI will own that clone. So the clone dies. All of the Settings UI is also destroyed, so I think that'll be fine.
</details>
I think we'll be fine. Would definitely like a second opinion from @DHowett if possible.

File: src/cascadia/TerminalApp/SettingsTab.cpp
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/SettingsTab.cpp:123]: Hmm... Maybe this context menu should be abstracted differently. I'm looking at #7728 and we would have to duplicate that code here and in `TerminalTab`. Here's a few ideas:

1. Could we make `ITab` an abstract class instead of an interface? That way the context menu gets inherited by both (and all future tabs)
2. Should we move the responsibility of populating the context menu over to `TerminalPage`? That way it checks the type of tab we have, and populates it accordingly?

File: src/cascadia/TerminalApp/SettingsTab.cpp
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/SettingsTab.cpp:176]: These are two other functions that are probably the same across both `ITab` implementations. Could we consolidate them somehow?

**Review by carlos-zamora — COMMENTED** [review]
This looks great! I want to download this branch and play around with it for a bit. I also have a few questions:
- Is the "Settings" button in the dropdown hookup up? If not, that should be an easy one we can cross of our list.

@zadjii-msft (for when you come back) I'd feel more comfortable if you took a closer look at this PR since you seem pretty familiar with this layer.

NOTE: this targets the feature branch so we're on a "1 approval policy". But we'll do one big review as a team when the Settings UI is ready and the feature branch gets merged into master.

File: src/cascadia/TerminalApp/TerminalAppLib.vcxproj
Comment (leonMSFT) [reply src/cascadia/TerminalApp/TerminalAppLib.vcxproj:30]: ah something keeps automagically keeps removing the spaces for me and it's on me to double check that I'm not pushing these in - maybe it's runformat? I'll revert it back 👍

Comment (carlos-zamora) [reply src/cascadia/TerminalApp/TerminalAppLib.vcxproj:30]: I think I ran into that problem before too, but I forget how. Nowadays, I just edit it in VS Code. Then when I switch to VS I get a prompt to reload all. That might work? idk 🤷‍♂️

File: src/cascadia/TerminalApp/TerminalTab.h
Comment (leonMSFT) [reply src/cascadia/TerminalApp/TerminalTab.h:97]: Instead of keeping the `Focus` as a boolean, I chose to copy the way XAML controls handles focus with `FocusStates` and bring it to `ITab` and all its implementers because I felt it would be good in general to know how the tab content is receiving focus. 

For example, having the ability to tell the Settings UI tab how it was focused (through a pointer interaction vs keyboard) could let us determine if we want to show a focus indicator outline and other a11y functionality as focus is gained. 

Unfortunately, we currently don't really plumb through our code the FocusState that started an action. That's why I've left all the focus calls as `Focus(FocusState::Programmatic)`.

File: src/cascadia/TerminalApp/TerminalPage.cpp
Comment (leonMSFT) [reply src/cascadia/TerminalApp/TerminalPage.cpp:2570]: exactly 😢

File: src/cascadia/TerminalApp/AppLogic.h
Comment (leonMSFT) [reply src/cascadia/TerminalApp/AppLogic.h:8]: hrmm you're right I don't think including `Tab` was needed in the first place

File: src/cascadia/TerminalApp/TerminalPage.cpp
Comment (leonMSFT) [reply src/cascadia/TerminalApp/TerminalPage.cpp:2625]: o good call, it's a one liner so I'll just replace this function call with that

File: src/cascadia/TerminalApp/TerminalPage.cpp
Comment (leonMSFT) [reply src/cascadia/TerminalApp/TerminalPage.cpp:2581]: Fine with me, I think it's only used here and in the function that creates a TerminalTab anyway

File: src/cascadia/TerminalApp/SettingsTab.cpp
Comment (leonMSFT) [reply src/cascadia/TerminalApp/SettingsTab.cpp:123]: I def agree with wanting to clean up the context menu creation in light of incoming additions in that PR and in the future. Here's another idea - I kinda want to make our own custom control named something like "TabContextFlyout". 
- Flyout menu items become xaml code - we won't need to have cpp code that creates our menu items. 
- Each tab type owns one of these controls.
- Each tab can hook itself up to the `Click` events of each of the control's menu items that they wish to use. Then we can turn each menu item's visibility off if the click handler isn't registered.

Comment (carlos-zamora) [reply src/cascadia/TerminalApp/SettingsTab.cpp:123]: You got me at "become xaml code" haha.

Comment (leonMSFT) [reply src/cascadia/TerminalApp/SettingsTab.cpp:111]: Alright this might take a little longer than I thought. I'm gonna file a follow up issue for this so that I can get this PR in, work on the issue on the side and unblock myself for some data binding work.

File: src/cascadia/TerminalApp/SettingsTab.cpp
Comment (leonMSFT) [reply src/cascadia/TerminalApp/SettingsTab.cpp:164]: Hmm, `_MakeSwitchToTabCommand` I can put inside of `TerminalPage`, the other function should actually be modified so that the setter for `TabViewIndex` also sets `SwitchToTabCommand`'s tabIndex.

File: src/cascadia/TerminalApp/SettingsTab.cpp
Comment (leonMSFT) [reply src/cascadia/TerminalApp/SettingsTab.cpp:110]: issue filed here: #7841

File: src/cascadia/TerminalApp/TerminalPage.cpp
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/TerminalPage.cpp:713]: Double checking. So we don't need this anymore?

File: src/cascadia/TerminalApp/TerminalPage.cpp
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/TerminalPage.cpp:2622]: Bare TODO

File: src/cascadia/TerminalApp/Utils.h
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/Utils.h:136]: If you haven't done this yet, make sure it's on the epic.

**Review by carlos-zamora — APPROVED** [review]
Approving w/ suggestions

File: src/cascadia/TerminalApp/TerminalPage.cpp
Comment (DHowett) [reply src/cascadia/TerminalApp/TerminalPage.cpp:2622]: This is my todo, you do not need to tag it.

File: src/cascadia/TerminalApp/TerminalPage.cpp
Comment (leonMSFT) [reply src/cascadia/TerminalApp/TerminalPage.cpp:713]: nope, it was redundant, the title change comes through the `PropertyChanged` handler

**Comment by german-one** [comment]
Ignore me if this was on purpose … The PR has not been merged into master.

**Comment by zadjii-msft** [comment]
Hey uh, was this supposed to get merged? since it wasn't going into `master`, the bot wasn't going to enforce the two signoff requirement...

**Comment by leonMSFT** [comment]
Oh yeah this was 100% intended to go into the feature branch for the SUI, _not_ master. I got so scared waking up to these two comments, I thought it accidentally went into master instead 😆

**Comment by german-one** [comment]
To ruin your morning was 100% unintended 😦

**Comment by carlos-zamora** [comment]
We're just doing one sign-off needed for feature/settings-ui. We'll do a full review when that merges into master in preparation for release. So occasionally we'll tag team members for a closer look or some immediate feedback. We were all surprised that `AutoMerge` worked hahaha.

**Comment by zadjii-msft** [comment]
Ah okay. I was planning on giving feedback on this part, but nbd. It would have probably boiled down to "I have concerns, but we need to ship the SUI so :shipit:". The concerns aren't that important now TBH

**Comment by DHowett** [comment]
@zadjii-msft they’re still worth tracking, so do write them up! Since this will merge to main eventually

**Pull request merged**
